### PR TITLE
Bluetooth: GATT: Fix race condition on requests

### LINKED ISSUE #17534: Race condition in GATT API.
Race-condition in the GATT api because att request information is
stored in user-space parameter structures. If the parameter is used
before the att request has been completed, the request structure will
be corrupted causing memory corruption.

**To Reproduce**
Unsubscribe directly after subscribing
``` 
static u8_t discover_func(struct bt_conn *conn,
@@ -88,6 +88,13 @@ static u8_t discover_func(struct bt_conn *conn,
                        printk("[SUBSCRIBED]\n");
                }
 
+               err = bt_gatt_unsubscribe(conn, &subscribe_params);
+               if (err && err != -EALREADY) {
+                       printk("Unsubscribe failed (err %d)\n", err);
+               } else {
+                       printk("[UNSUBSCRIBED]\n");
+               }
+
                return BT_GATT_ITER_STOP;
        }
```

**Expected behavior**
API should succeed, or return error code.

**Impact**
System crash.

**Screenshots or console output**
```
[00:00:00.006,469] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
[00:00:00.006,500] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 1.14 Build 9
[00:00:00.007,049] <inf> bt_hci_core: Identity: c1:f6:39:6f:90:c5 (random)
[00:00:00.007,049] <inf> bt_hci_core: HCI: version 5.0 (0x09) revision 0x0000, manufacturer 0x05f1
[00:00:00.007,049] <inf> bt_hci_core: LMP: version 5.0 (0x09) subver 0xffff
[ATTRIBUTE] handle 19
[ATTRIBUTE] handle 21
[SUBSCRIBED]
[UNSUBSCRIBED]
[UNSUBSCRIBED]
***** MPU FAULT *****
  Data Access Violation
  MMFAR Address: 0x4
***** Hardware exception *****
Current thread ID = 0x200017e0
Faulting instruction address = 0x1a7c8
Fatal fault in thread 0x200017e0! Aborting.
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK,
 - Commit: 8083f8a6b06751be7710e42b0ca370949825b408

### PR DESCRIPTION
**Pull request by Vudentz**
This fixes a race condition where an outstanding request could be canceled and replaced with another request breaking ATT flow and causing memory corruptions.

Fixes #17534

### CODE REVIEW

**Review by joerchan — COMMENTED** [review]
Should we try to do something if the user mistankely issues bt_gatt_write with the same parameters before the write is finished? I think the same problem of re-using the att_req struct exists then. But that seems more like a failure of the user to use the API properly.

File: subsys/bluetooth/host/att.c [hunk 1]
```diff
@@ -2318,7 +2326,7 @@ void bt_att_req_cancel(struct bt_conn *conn, struct bt_att_req *req)
 
 	/* Check if request is outstanding */
 	if (att->req == req) {
-		att->req = NULL;
+		att->req = &cancel;
```
[the remark this answers is not in the document]
Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: That is cleared later by att_req_detroy.

File: subsys/bluetooth/host/gatt.c [hunk 2]
```diff
@@ -3013,6 +3013,8 @@ static void gatt_write_ccc_rsp(struct bt_conn *conn, u8_t err,
 
 	BT_DBG("err 0x%02x", err);
 
+	params->flags &= ~BT_GATT_SUBSCRIBE_WRITE_PENDING;
```
[the remark this answers is not in the document]
Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: Yep, though flags are not atomic_t but I guess we can change that to make it simpler for doing this.

**Comment by Vudentz** [comment]
> Should we try to do something if the user mistankely issues bt_gatt_write with the same parameters before the write is finished? I think the same problem of re-using the att_req struct exists then. But that seems more like a failure of the user to use the API properly.

Yep, Id said that is improper use of the API and it is stated on the documentation, that said Im open to change the GATT client APIs in the future to make it simpler to enqueue more requests if there is memory available.

**Comment by Vudentz** [comment]
@joerchan I added the change to use atomic_t for subscribe flags, I guess this one needs to be market for 1.14 as well.

File: include/kernel.h [hunk 3]
```diff
@@ -4248,7 +4248,7 @@ extern void k_mem_pool_free_id(struct k_mem_block_id *id);
  * allocated from the heap memory pool.
  *
  * @param size Amount of memory requested (in bytes).
- *
+ id*
```
Comment (joerchan) [inline include/kernel.h]: Accident?

File: subsys/bluetooth/host/gatt.c [hunk 4]
```diff
@@ -3050,6 +3053,8 @@ static int gatt_write_ccc(struct bt_conn *conn, u16_t handle, u16_t value,
 
 	BT_DBG("handle 0x%04x value 0x%04x", handle, value);
 
+	atomic_set_bit(&params->flags, BT_GATT_SUBSCRIBE_WRITE_PENDING);
```
Comment (joerchan) [inline subsys/bluetooth/host/gatt.c]: Drop the &

Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: Should be fixed now

File: subsys/bluetooth/host/gatt.c [hunk 5]
```diff
@@ -3013,6 +3014,8 @@ static void gatt_write_ccc_rsp(struct bt_conn *conn, u8_t err,
 
 	BT_DBG("err 0x%02x", err);
 
+	atomic_clear_bit(params->flags, BT_GATT_SUBSCRIBE_NUM_FLAGS);
```
[the remark this answers is not in the document]
Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: Fixed.

**Review by joerchan — APPROVED** [review]

**Comment by Vudentz** [comment]
@carlescufi can we get this merged?

**Review by carlescufi — APPROVED** [review]

**Pull request merged**